activation/activation.go

```diff
@@ -914,54 +900,38 @@ func (b *Builder) broadcast(ctx context.Context, atx scale.Encodable) (int, erro
 	return len(buf), nil
 }
 
-// searchPositioningAtx returns atx id with the highest tick height.
-// Publish epoch is used for caching the positioning atx.
-func (b *Builder) searchPositioningAtx(
+// find returns atx id with the highest tick height.
+// The publish epoch (of the built ATX) is used for:
+// - caching the positioning atx,
+// - filtering candidates for positioning atx (it must be published in an earlier epoch than built ATX).
+//
+// It always returns an ATX, falling back to the golden one as the last resort.
+func (f *positioningAtxFinder) find(
 	ctx context.Context,
-	nodeID types.NodeID,
+	atxs AtxService,
 	publish types.EpochID,
-) (types.ATXID, error) {
-	logger := b.logger.With(log.ZShortStringer("smesherID", nodeID), zap.Uint32("publish epoch", publish.Uint32()))
+) types.ATXID {
+	logger := f.logger.With(zap.Uint32("publish epoch", publish.Uint32()))
 
-	b.posAtxFinder.finding.Lock()
-	defer b.posAtxFinder.finding.Unlock()
+	f.finding.Lock()
+	defer f.finding.Unlock()
 
-	if found := b.posAtxFinder.found; found != nil && found.forPublish == publish {
+	if found := f.found; found != nil && found.forPublish == publish {
 		logger.Debug("using cached positioning atx", log.ZShortStringer("atx_id", found.id))
-		return found.id, nil
-	}
-
-	latestPublished, err := atxs.LatestEpoch(b.db)
-	if err != nil {
-		return types.EmptyATXID, fmt.Errorf("get latest epoch: %w", err)
+		return found.id
 	}
 
-	logger.Info("searching for positioning atx", zap.Uint32("latest_epoch", latestPublished.Uint32()))
-
-	// positioning ATX publish epoch must be lower than the publish epoch of built ATX
-	positioningAtxPublished := min(latestPublished, publish-1)
-	id, err := findFullyValidHighTickAtx(
-		ctx,
-		b.atxsdata,
-		positioningAtxPublished,
-		b.conf.GoldenATXID,
-		b.validator,
-		logger,
-		VerifyChainOpts.AssumeValidBefore(time.Now().Add(-b.postValidityDelay)),
-		VerifyChainOpts.WithTrustedID(nodeID),
-		VerifyChainOpts.WithLogger(b.logger),
-	)
+	id, err := atxs.PositioningATX(ctx, publish-1)
 	if err != nil {
-		logger.Info("search failed - using golden atx as positioning atx", zap.Error(err))
-		id = b.conf.GoldenATXID
+		logger.Warn("failed to get positioning ATX - falling back to golden", zap.Error(err))
+		f.found = &foundPosAtx{f.golden, publish}
+		return f.golden
```

Comment:
Again I think in this case we should return `types.EmptyATXID` and `ErrNotFound`. If we couldn't find an ATX the client / caller has to make a decision which ATX to use (golden or prevATX). To check if the returned ATX is the golden ATX to make that decision it already needs to know the ID of that ATX so returning a "not found" error seems like a better choice to me.

Reply:
Here, I want to persist picked positioning ATX already. There could be many calls from different IDs happening in parallel. The first one selects the ATX (which is a long, costly process) while the rest is blocked on the mutex. As soon as the first one picks a value and returns - all the other use it. If it didn't fall back to the golden ATX here, all other calls would repeat costly searching (and most likely end up using the golden one anyway).

Reply:
Would it be possible to use singleflight instead of the mutex here instead? So if multiple clients wait for a positioning ATX and none could be found in time all waiting will receive "not found" and can then pick the golden ATX or their previous ATX.

The next call would then start another search (which in my opinion it should) to try to find an ATX again?

Wdyt?